File: cpi/signals/nasa_firms.py
# signals/nasa_firms.py — Satellite thermal anomaly detection (BINARY OVERRIDE)
# Not a CPI input. Fires as a scenario probability override when
# thermal anomalies are detected at critical infrastructure ABOVE baseline.
# Uses NASA FIRMS VIIRS data — free, no API key, updates every 3h.
#
# Key design: industrial regions (South Pars, refineries) have thermal
# signatures every day. We only alert when hits EXCEED the daily baseline.
# Bushehr Nuclear has baseline=0 — any hit there is significant.

import csv
import io
import math
import requests
from datetime import datetime, timezone

import sys, os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import (
    FIRMS_URL, FIRMS_URL_FALLBACK, FIRMS_TIMEOUT, FIRMS_MIN_FRP,
    FIRMS_PREFER_NIGHT, FIRMS_CRITICAL_REGIONS, FIRMS_KNOWN_FLARES
)


def _no_data(reason):
    """
    Missing satellite data ≠ "no strikes."
    alert=None (unknown), not alert=False (safe).
    """
    return {
        "signal": "nasa_firms", "type": "override",
        "alert": None,
        "confidence": 0.0,
        "total_hits": 0, "critical_hits": {}, "above_baseline": {},
        "interpretation": f"NO DATA: {reason} — cannot confirm or deny strikes",
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def _is_known_flare(lat, lon):
    """Check if point is near a known persistent flare / industrial site."""
    for flare_lat, flare_lon, radius in FIRMS_KNOWN_FLARES:
        dist = math.sqrt((lat - flare_lat)**2 + (lon - flare_lon)**2)
        if dist < radius:
            return True
    return False


def _in_region(lat, lon, bounds):
    """bounds = (lat_min, lat_max, lon_min, lon_max, baseline)"""
    lat_min, lat_max, lon_min, lon_max = bounds[:4]
    return lat_min <= lat <= lat_max and lon_min <= lon <= lon_max


def _fetch_firms_csv():
    """Fetch FIRMS global 24h CSV (~5MB). Nighttime + geo filters cut this to a handful."""
    print(f"[FIRMS] Downloading global CSV...")
    resp = requests.get(FIRMS_URL, timeout=FIRMS_TIMEOUT)
    resp.raise_for_status()
    print(f"[FIRMS] Downloaded: {len(resp.text):,} bytes")
    return resp.text


def collect_nasa_firms():
    try:
        csv_text = _fetch_firms_csv()
        reader = csv.DictReader(io.StringIO(csv_text))

        required_cols = {"latitude", "longitude", "frp", "confidence"}
        if reader.fieldnames and not required_cols.issubset(set(reader.fieldnames)):
            missing = required_cols - set(reader.fieldnames)
            return _no_data(f"CSV schema changed — missing columns: {missing}")

        total_hits = 0
        critical_hits = {}   # region_name → raw count
        flare_filtered = 0   # how many we skipped as known flares
        day_filtered = 0     # how many we skipped as daytime (sun glint risk)

        for row in reader:
            try:
                lat = float(row["latitude"])
                lon = float(row["longitude"])
                frp = float(row.get("frp", 0))
                conf = row.get("confidence", "l")
                daynight = row.get("daynight", "")
            except (ValueError, KeyError):
                continue

            # Skip daytime detections if configured — eliminates sun glint
            if FIRMS_PREFER_NIGHT and daynight == "D":
                day_filtered += 1
                continue

            # Skip low-confidence AND low-power detections
            if conf == "l" and frp < FIRMS_MIN_FRP:
                continue

            # Skip known persistent flare / industrial sites
            if _is_known_flare(lat, lon):
                flare_filtered += 1
                continue

            # Only count Iran/Gulf area (22-40N, 44-64E)
            # (redundant if using regional URL, but safe for global fallback)
            if not (22 <= lat <= 40 and 44 <= lon <= 64):
                continue

            total_hits += 1

            # Check critical infrastructure regions
            for region_name, bounds in FIRMS_CRITICAL_REGIONS.items():
                if _in_region(lat, lon, bounds):
                    critical_hits[region_name] = critical_hits.get(region_name, 0) + 1

        # ── Baseline-aware alerting ───────────────────────────────
        # Only alert if hits EXCEED the daily baseline for a region.
        # A gas field with baseline=5 needs 6+ hits to trigger.
        above_baseline = {}
        for region_name, count in critical_hits.items():
            bounds = FIRMS_CRITICAL_REGIONS[region_name]
            baseline = bounds[4] if len(bounds) > 4 else 0
            excess = count - baseline
            if excess > 0:
                above_baseline[region_name] = excess

        # Alert only on above-baseline critical hits or very high total
        alert = bool(above_baseline) or total_hits > 30

        # ── Interpretation ────────────────────────────────────────
        if above_baseline:
            parts = []
            for region, excess in above_baseline.items():
                raw = critical_hits[region]
                baseline = FIRMS_CRITICAL_REGIONS[region][4] if len(FIRMS_CRITICAL_REGIONS[region]) > 4 else 0
                parts.append(f"{region} ({raw} hits, baseline {baseline}, +{excess} above)")
            regions_str = ", ".join(parts)
            interp = f"ABOVE-BASELINE thermal at {regions_str} — possible strike"
        elif critical_hits:
            # Hits detected but within normal baseline
            parts = [f"{k} ({v})" for k, v in critical_hits.items()]
            regions_str = ", ".join(parts)
            interp = f"Thermal at {regions_str} — within industrial baseline (normal)"
        elif total_hits > 30:
            interp = f"{total_hits} thermal detections across Iran/Gulf — elevated"
        elif total_hits > 0:
            interp = f"{total_hits} thermal detections — within normal range"
        else:
            interp = "No significant thermal anomalies in Iran/Gulf region"

        # Log diagnostic info
        if day_filtered > 0:
            print(f"[FIRMS] Filtered {day_filtered} daytime detections (sun glint risk)")
        if flare_filtered > 0:
            print(f"[FIRMS] Filtered {flare_filtered} known flare detections")
        if critical_hits:
            for r, c in critical_hits.items():
                baseline = FIRMS_CRITICAL_REGIONS[r][4] if len(FIRMS_CRITICAL_REGIONS[r]) > 4 else 0
                status = "ABOVE" if c > baseline else "normal"
                print(f"[FIRMS] {r}: {c} hits (baseline={baseline}) → {status}")

        return {
            "signal": "nasa_firms", "type": "override",
            "alert": alert, "confidence": 1.0,
            "total_hits": total_hits,
            "critical_hits": critical_hits,
            "above_baseline": above_baseline,
            "flare_filtered": flare_filtered,
            "interpretation": interp,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }

    except requests.exceptions.Timeout:
        return _no_data(f"FIRMS CSV timeout after {FIRMS_TIMEOUT}s")
    except Exception as e:
        return _no_data(f"FIRMS error — {str(e)[:80]}")

File: src/lib/cpi/signals/nasa-firms.ts
// signals/nasa-firms.ts — Satellite thermal anomaly detection (BINARY OVERRIDE)
// Uses NASA FIRMS VIIRS data. Nighttime-only + geo filter + known flare filter.
// Baseline-aware: only alerts when hits EXCEED daily baseline per region.

import type { SignalResult } from "../types";
import {
  FIRMS_URL,
  FIRMS_TIMEOUT,
  FIRMS_MIN_FRP,
  FIRMS_PREFER_NIGHT,
  FIRMS_CRITICAL_REGIONS,
  FIRMS_KNOWN_FLARES,
} from "../config";

function noData(reason: string): SignalResult {
  return {
    signal: "nasa_firms",
    type: "override",
    score: 50,
    alert: null,
    confidence: 0.0,
    total_hits: 0,
    critical_hits: {},
    above_baseline: {},
    interpretation: `NO DATA: ${reason} \u2014 cannot confirm or deny strikes`,
    timestamp: new Date().toISOString(),
  };
}

function isKnownFlare(lat: number, lon: number): boolean {
  for (const [fLat, fLon, radius] of FIRMS_KNOWN_FLARES) {
    const dist = Math.sqrt((lat - fLat) ** 2 + (lon - fLon) ** 2);
    if (dist < radius) return true;
  }
  return false;
}

function inRegion(
  lat: number,
  lon: number,
  bounds: [number, number, number, number, number]
): boolean {
  return (
    lat >= bounds[0] &&
    lat <= bounds[1] &&
    lon >= bounds[2] &&
    lon <= bounds[3]
  );
}

// Simple CSV parser for FIRMS data (no external dependency needed)
function parseCSV(text: string): Record<string, string>[] {
  const lines = text.split("\n");
  if (lines.length < 2) return [];
  const headers = lines[0].split(",").map((h) => h.trim());
  const rows: Record<string, string>[] = [];
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    const values = line.split(",");
    const row: Record<string, string> = {};
    for (let j = 0; j < headers.length; j++) {
      row[headers[j]] = (values[j] ?? "").trim();
    }
    rows.push(row);
  }
  return rows;
}

export async function collectNasaFirms(): Promise<SignalResult> {
  try {
    console.log("[FIRMS] Downloading global CSV...");
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), FIRMS_TIMEOUT);

    const resp = await fetch(FIRMS_URL, { signal: controller.signal });
    clearTimeout(timer);

    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const csvText = await resp.text();
    console.log(`[FIRMS] Downloaded: ${csvText.length.toLocaleString()} bytes`);

    const rows = parseCSV(csvText);
    if (rows.length === 0) return noData("empty CSV");

    // Verify columns
    const firstRow = rows[0];
    for (const col of ["latitude", "longitude", "frp", "confidence"]) {
      if (!(col in firstRow)) {
        return noData(`CSV schema changed \u2014 missing column: ${col}`);
      }
    }

    let totalHits = 0;
    const criticalHits: Record<string, number> = {};
    let flareFiltered = 0;
    let dayFiltered = 0;

    for (const row of rows) {
      const lat = parseFloat(row.latitude);
      const lon = parseFloat(row.longitude);
      const frp = parseFloat(row.frp || "0");
      const conf = row.confidence || "l";
      const daynight = row.daynight || "";

      if (isNaN(lat) || isNaN(lon)) continue;

      // Skip daytime (sun glint risk)
      if (FIRMS_PREFER_NIGHT && daynight === "D") {
        dayFiltered++;
        continue;
      }

      // Skip low-confidence + low-power
      if (conf === "l" && frp < FIRMS_MIN_FRP) continue;

      // Skip known flares
      if (isKnownFlare(lat, lon)) {
        flareFiltered++;
        continue;
      }

      // Only Iran/Gulf area
      if (!(lat >= 22 && lat <= 40 && lon >= 44 && lon <= 64)) continue;

      totalHits++;

      // Check critical regions
      for (const [regionName, bounds] of Object.entries(
        FIRMS_CRITICAL_REGIONS
      )) {
        if (inRegion(lat, lon, bounds)) {
          criticalHits[regionName] =
            (criticalHits[regionName] ?? 0) + 1;
        }
      }
    }

    // Baseline-aware alerting
    const aboveBaseline: Record<string, number> = {};
    for (const [regionName, count] of Object.entries(criticalHits)) {
      const bounds = FIRMS_CRITICAL_REGIONS[regionName];
      const baseline = bounds[4];
      const excess = count - baseline;
      if (excess > 0) aboveBaseline[regionName] = excess;
    }

    const alert =
      Object.keys(aboveBaseline).length > 0 || totalHits > 30;

    // Interpretation
    let interp: string;
    if (Object.keys(aboveBaseline).length > 0) {
      const parts = Object.entries(aboveBaseline).map(
        ([region, excess]) => {
          const raw = criticalHits[region];
          const baseline = FIRMS_CRITICAL_REGIONS[region][4];
          return `${region} (${raw} hits, baseline ${baseline}, +${excess} above)`;
        }
      );
      interp = `ABOVE-BASELINE thermal at ${parts.join(", ")} \u2014 possible strike`;
    } else if (Object.keys(criticalHits).length > 0) {
      const parts = Object.entries(criticalHits).map(
        ([k, v]) => `${k} (${v})`
      );
      interp = `Thermal at ${parts.join(", ")} \u2014 within industrial baseline (normal)`;
    } else if (totalHits > 30) {
      interp = `${totalHits} thermal detections across Iran/Gulf \u2014 elevated`;
    } else if (totalHits > 0) {
      interp = `${totalHits} thermal detections \u2014 within normal range`;
    } else {
      interp = "No significant thermal anomalies in Iran/Gulf region";
    }

    // Log diagnostics
    if (dayFiltered > 0)
      console.log(
        `[FIRMS] Filtered ${dayFiltered} daytime detections`
      );
    if (flareFiltered > 0)
      console.log(
        `[FIRMS] Filtered ${flareFiltered} known flare detections`
      );
    for (const [r, c] of Object.entries(criticalHits)) {
      const baseline = FIRMS_CRITICAL_REGIONS[r][4];
      const status = c > baseline ? "ABOVE" : "normal";
      console.log(
        `[FIRMS] ${r}: ${c} hits (baseline=${baseline}) -> ${status}`
      );
    }

    return {
      signal: "nasa_firms",
      type: "override",
      score: 50,
      alert,
      confidence: 1.0,
      total_hits: totalHits,
      critical_hits: criticalHits,
      above_baseline: aboveBaseline,
      flare_filtered: flareFiltered,
      interpretation: interp,
      timestamp: new Date().toISOString(),
    };
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    if (msg.includes("abort")) {
      return noData(`FIRMS CSV timeout after ${FIRMS_TIMEOUT / 1000}s`);
    }
    return noData(`FIRMS error \u2014 ${msg.slice(0, 80)}`);
  }
}
